src/limiter/bound_preserving_limiter.cpp

```diff
@@ -0,0 +1,1003 @@
+            dealii::LinearAlgebra::distributed::Vector<double>      solution,
+            const dealii::DoFHandler<dim>&                          dof_handler,
+            const dealii::hp::FECollection<dim>&                    fe_collection)
+        {
+            for (auto soln_cell : dof_handler.active_cell_iterators()) {
+                if (!soln_cell->is_locally_owned()) continue;
+
+                std::vector<dealii::types::global_dof_index> current_dofs_indices;
+                // Current reference element related to this physical cell
+                const int i_fele = soln_cell->active_fe_index();
+                const int poly_degree = i_fele;
+
+                const dealii::FESystem<dim, dim>& current_fe_ref = fe_collection[poly_degree];
+                const unsigned int n_dofs_curr_cell = current_fe_ref.n_dofs_per_cell();
+                // Obtain the mapping from local dof indices to global dof indices
+                current_dofs_indices.resize(n_dofs_curr_cell);
+                soln_cell->get_dof_indices(current_dofs_indices);
+
+                // Extract the local solution dofs in the cell from the global solution dofs
+                if (this->global_max.size() < nstate && this->global_min.size() < nstate) {
+                    for (unsigned int istate = 0; istate < nstate; ++istate) {
+                        this->global_max.push_back(-1e9);
+                        this->global_min.push_back(1e9);
+                    }
+                }
+
+                // Allocate solution dofs and set global max and min
+                std::array<std::vector<real>, nstate> soln_dofs;
+                const unsigned int n_shape_fns = n_dofs_curr_cell / nstate;
+                for (unsigned int idof = 0; idof < n_dofs_curr_cell; ++idof) {
+                    const unsigned int istate = fe_collection[poly_degree].system_to_component_index(idof).first;
+                    const unsigned int ishape = fe_collection[poly_degree].system_to_component_index(idof).second;
+                    if (ishape == 0) {
+                        soln_dofs[istate].resize(n_shape_fns);
+                    }
+                    soln_dofs[istate][ishape] = solution[current_dofs_indices[idof]]; //
+                    if (soln_dofs[istate][ishape] > this->global_max[istate])
+                        this->global_max[istate] = soln_dofs[istate][ishape];
+                    if (soln_dofs[istate][ishape] < this->global_min[istate])
+                        this->global_min[istate] = soln_dofs[istate][ishape];
+                }
+            }
+
+            // Print the obtained values for verification
+            for (unsigned int istate = 0; istate < nstate; ++istate) {
+                std::cout << std::fixed;
+                std::cout << std::setprecision(14);
+                std::cout << "global_max:   " << this->global_max[istate] << "   global_min:   " << this->global_min[istate] << std::endl;
```

Comment:
The global maximum and minimum are maximum and minimum of the initial condition, so the function is only called once at the beginning of the test. I don't currently have a case where it would have to be called multiple times, but if there is, I can definitely include this as a future item!

Reply:
Sounds reasonable!